src/Microsoft.Windows.ComputeVirtualization/HcsNotification.cs

```diff
@@ -0,0 +1,100 @@
+using System;
+using System.Runtime.InteropServices;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Windows.ComputeVirtualization
+{
+    internal class NotificationResult
+    {
+        public int Status;
+        public string Data;
+    }
+
+    internal enum HCS_NOTIFICATIONS : uint
+    {
+        HcsNotificationInvalid = 0x00000000,
+
+        /// Notifications for HCS_SYSTEM handles
+        HcsNotificationSystemExited = 0x00000001,
+        HcsNotificationSystemCreateCompleted = 0x00000002,
+        HcsNotificationSystemStartCompleted = 0x00000003,
+
+        /// Notifications for HCS_PROCESS handles
+        HcsNotificationProcessExited = 0x00010000,
+
+        /// Common notifications
+        HcsNotificationServiceDisconnect = 0x01000000,
+
+        /// The upper 4 bits are reserved for flags. See HCS_NOTIFICATION_FLAGS
+        HcsNotificationFlagsReserved = 0xF0000000
+    }
+
+    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
+    internal delegate void NotificationCallback(uint notificationType, IntPtr context, int notificationStatus, [MarshalAs(UnmanagedType.LPWStr)] string notificationData);
+
+    internal delegate int RegisterHcsNotificationCallback(IntPtr handle, NotificationCallback callback, IntPtr context, out IntPtr callbackHandle);
+
+    internal delegate int UnregisterHcsNotificationCallback(IntPtr handle);
+
+    internal class HcsNotificationWatcher : IDisposable
+    {
+        private IntPtr _h;
+        private UnregisterHcsNotificationCallback _unreg;
+        private readonly Dictionary<HCS_NOTIFICATIONS, TaskCompletionSource<NotificationResult>> _n = new Dictionary<HCS_NOTIFICATIONS, TaskCompletionSource<NotificationResult>>();
+
+        public HcsNotificationWatcher(IntPtr handle, RegisterHcsNotificationCallback register, UnregisterHcsNotificationCallback unregister, HCS_NOTIFICATIONS[] notificationList)
+        {
+            _unreg = unregister;
+
+            foreach (var notificationType in notificationList)
+            {
+                var entry = new TaskCompletionSource<NotificationResult>();
```

Comment:
I think this pattern `only` works because you haven't ever tried to wait for the same event twice. You would need a new completion source. Is this a possible? I have a hard time thinking of the workflow where I would wait again but it seems plausible.